zcash_client_sqlite/src/wallet/transparent.rs

```diff
@@ -995,21 +995,52 @@ pub(crate) fn tx_unexpired_condition_minconf_0(tx: &str) -> String {
+/// to be unspent after the transaction that created it would have expired.
+///
+/// TODO: This fragment is very unwieldy; it really doesn't work well as a fragment. It would be
+/// much better if it could be expressed as a view somehow; the problem is that views can't be
+/// parameterized, and there are numerous interacting checks that need to be made against the
+/// target height in the context where this fragment is being used.
 ///
 /// # Usage requirements
-/// - `tx` must be set to the alias for the `transactions` table in the enclosing scope.
-/// - `accounts` must be set to the alias for the `accounts` table in the enclosing scope.
+/// - `transparent_received_outputs` must be set to the alias to the `transparent_received_outputs`
+///   table in the enclosing scope.
+/// - `addresses` must be set to alias for the `addresses` table in the enclosing scope such that
+///   `addresses.id = transparent_received_outputs.address_id`.
+/// - `tx` must be set to the alias for the `transactions` table in the enclosing scope such that
+///   `tx.id_tx = transparent_received_outputs.transaction_id`.
+/// - `accounts` must be set to the alias for the `accounts` table in the enclosing scope such that
+///   `accounts.id = transparent_received_outputs.account_id`
 /// - The parent is responsible for enclosing this condition in parentheses as appropriate.
-pub(crate) fn tx_spends_account_outputs_condition(tx: &str, accounts: &str) -> String {
+/// - The parent is responsible for ensuring that this condition will only be checked for
+///   outputs that have already otherwise been verified to be spendible, i.e. it must be
+///   used as a strictly constricting clause on the set of outputs.
+pub(crate) fn excluding_wallet_internal_ephemeral_outputs(
+    transparent_received_outputs: &str,
+    addresses: &str,
+    tx: &str,
+    accounts: &str,
+) -> String {
+    let ephemeral_key_scope = KeyScope::Ephemeral.encode();
     format!(
         r#"
-        {tx}.id_tx IN (
+        -- the receiving address is not an ephemeral address
+        {addresses}.key_scope != {ephemeral_key_scope}
+        -- or the transaction that generated the TXO has no inputs belonging to the wallet
+        OR {tx}.id_tx NOT IN (
             SELECT transaction_id
             FROM v_received_output_spends
             WHERE v_received_output_spends.account_id = {accounts}.id
         )
+        -- or the transaction that generated the TXO would be considered expired as of the TXOs
+        -- max_observed_unspent_height; this operates under the assumption that the second
+        -- transaction in a TEX chain has the same expiry as the transaction that generated the
+        -- ephemeral output, and the output having been observed to be unspent above this height
+        -- indicates that the subsequent spend failed and the spending transaction will have
+        -- expired.
+        OR {transparent_received_outputs}.max_observed_unspent_height > {tx}.expiry_height
```

Comment:
Bug: this should be
```suggestion
        OR {transparent_received_outputs}.max_observed_unspent_height >= {tx}.expiry_height
```
because the expiry height is the last valid height the tx can be mined at, so if the tx is observed unspent as of the end of the block at that height (which is what `max_observed_unspent_height` records), then the tx cannot be mined.

Add a test that catches this bug by exercising the boundary.

Reply:
This isn't a bug; using `>` instead of `>=` is the more conservative choice: we only allow the TXO to be selected once it has been observed to be unspent at a height _greater_ than the expiry height.

In fact, I think your suggested change here would be a bug. Note that the sense of this whole clause has changed, I rewrote it via De Morgan's law in order to be able to include this third clause via an `OR`.

Reply:
My brain was misparsing `unspent` as `unmined`, which led to my mis-analysis.